moonraker/components/sensor.py

```diff
@@ -0,0 +1,232 @@
+        err: Optional[Exception] = None
+        context = {"payload": payload.decode()}
+        logging.debug("Context: %s", context)
+        try:
+            response = float(self.state_response.render(context))
+        except Exception as e:
+            self.error_state = str(e)
+        else:
+            self.error_state = None
+
+            self.last_value = round(response, self.config.accuracy_decimals)
+            logging.debug(
+                "Received updated sensor value for %s: %s%s",
+                self.config.name,
+                self.last_value,
+                self.config.unit,
+            )
+
+    async def _on_mqtt_disconnected(self):
+        self.error_state = "MQTT Disconnected"
+        self.last_value = 0.0
+
+    async def initialize(self) -> None:
+        try:
+            self.mqtt.subscribe_topic(
+                self.state_topic, self._on_state_update, self.qos
+            )
+            self.error_state = None
+        except Exception as e:
+            self.error_state = str(e)
+
+        return await super().initialize()
+
+
+class Sensors:
+    __sensor_types: Dict[str, Type[BaseSensor]] = {"MQTT": MQTTSensor}
+
+    def __init__(self, config: ConfigHelper) -> None:
+        self.server = config.get_server()
+        prefix_sections = config.get_prefix_sections("sensor")
+
+        self.sensors = {}
+        # Register endpoints
+        self.server.register_endpoint(
+            "/server/sensors", ["GET"], self._handle_sensor_data_request
+        )
+
+        for section in prefix_sections:
+            cfg = config[section]
+
+            try:
+                try:
+                    _, name = cfg.get_name().split(maxsplit=1)
+                except ValueError:
+                    raise cfg.error(f"Invalid section name: {cfg.get_name()}")
+
+                logging.info(f"Configuring sensor: {name}")
+
+                sensor_type: str = cfg.get("type", "mqtt")
+                sensor_class: Optional[
+                    Type[BaseSensor]
+                ] = self.__sensor_types.get(sensor_type.upper(), None)
+                if sensor_class is None:
+                    raise config.error(
+                        f"Unsupported sensor type: {sensor_type}"
+                    )
+
+                self.sensors[name] = sensor_class(name=name, cfg=cfg)
+            except Exception as e:
+                # Ensures that configuration errors are shown to the user
+                self.server.add_warning(
+                    f"Failed to configure sensor [{cfg.get_name()}]\n{e}"
+                )
+                continue
+
+    async def component_init(self) -> None:
+        try:
+            logging.debug("Initializing sensor component")
+            event_loop = self.server.get_event_loop()
+            cur_time = event_loop.get_loop_time()
+            endtime = cur_time + 120.0
+
+            uninitialized_sensors = list(self.sensors.values())
+            while cur_time < endtime:
```

Comment:
This logic was taken from the `power` module before it was refactored.  There is no longer a loop in `component_init`, instead each sensor's init routine is executed asynchronously.  This init procedure is indefinite, however it does log failures.  This was done for two reasons:

1) The server won't start until `component_init` completes for each component.  We want to avoid delaying it.
2) When the machine first boots this can cause failures because the network isn't up yet.

That said, the `mqtt` implementation of initialize doesn't event need to be async.  It doesn't await anything and init occurs when it the subscription data is received.  The above will only be necessary for implementations that require polling should they be added.